Objects/cown.c

```diff
@@ -0,0 +1,323 @@
+    }
+
+    if (value) {
+        PyCown_set_unchecked(self, value);
+
+    } else {
+        atomic_store(&self->state, Cown_RELEASED);
+        self->value = Py_None;
+    }
+    return 0;
+}
+
+static int PyCown_traverse(PyCownObject *self, visitproc visit, void *arg) {
+    Py_VISIT(self->value);
+    return 0;
+}
+
+#define bail_if_owned(o, msg) \
+    do { \
+        /* Note: we must hold the GIL at this point -- note for future threading implementation. */ \
+        if (o->owning_thread != 0) { \
+            PyErr_Format(PyExc_RegionError, "%s: %S -- %zd", msg, o, o->owning_thread); \
+            return NULL; \
+        } \
+    } while(0);
+
+#define bail_unless_owned(o, msg) \
+    do { \
+        /* Note: we must hold the GIL at this point -- note for future threading implementation. */ \
+        PyThreadState *tstate = PyThreadState_Get(); \
+        if (o->owning_thread != tstate->thread_id) { \
+            PyErr_Format(PyExc_RegionError, "%s: %S", msg, o); \
+            return NULL; \
+        } \
+    } while(0);
+
+#define bail_unless_in_state(o, expected_state, msg) \
+    do { \
+        /* Note: we must hold the GIL at this point -- note for future threading implementation. */ \
+        if (o->state != expected_state) { \
+            PyErr_Format(PyExc_RegionError, "%s: %S", msg, o); \
+            return NULL; \
+        } \
+    } while(0);
+
+#define bail_unless_acquired(o, msg) \
+    bail_unless_owned(o, msg) \
+    bail_unless_in_state(o, Cown_ACQUIRED, msg)
+
+static PyObject *PyCown_acquire(PyCownObject *self, PyObject *args) {
+    int expected = 0;
+
+    while (!atomic_compare_exchange_strong(&self->state, &expected, Cown_ACQUIRED)) {
+        sem_wait(&self->semaphore);
+        expected = 0;
+    }
+
+    // Note: we must hold the GIL at this point -- note for future
+    // threading implementation.
+    PyThreadState *tstate = PyThreadState_Get();
+    self->owning_thread = tstate->thread_id;
+
+#ifndef NDEBUG
+    fprintf(stderr, "Cown %p acquired by %zd\n", self, tstate->thread_id);
+#endif
+    Py_RETURN_NONE;
+}
+
+static PyObject *PyCown_release(PyCownObject *self, PyObject *args) {
+    if (self->state == Cown_RELEASED) {
+        bail_if_owned(self, "BUG: Released cown had owning thread: %p");
+        Py_RETURN_NONE;
+    }
+
+    bail_unless_owned(self, "Thread attempted to release a cown it did not own");
+
+    atomic_store(&self->state, Cown_RELEASED);
+    self->owning_thread = 0;
+    sem_post(&self->semaphore);
+
+#ifndef NDEBUG
+    fprintf(stderr, "Cown %p released\n", self);
+#endif
+    Py_RETURN_NONE;
+}
+
+int _PyCown_release(PyObject *self) {
+    PyObject* res = PyCown_release((PyCownObject *)self, NULL);
+    return res == Py_None ? 0 : -1;
+}
+
+int _PyCown_is_released(PyObject *self) {
+    PyCownObject *cown = (PyCownObject *)self;
+    return cown->state == Cown_RELEASED;
+}
+
+static PyObject *PyCown_get(PyCownObject *self, PyObject *args) {
```

Comment:
What is `args` for here?

Reply:
At some point I believed that I had to have these extra arguments to avoid a stupid warning in C but that seems to not be the case. Thanks!